mautic/mautic
# Adds option to set custom headers for emails

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds support to set custom headers on outgoing emails configurable through the advanced settings of an email. It also adds custom headers for non-Email entity related sends to the system configuration. System configuration headers are not auto-added to any email based on the Email entity (campaign email, broadcast email, etc). System configuration headers apply to emails such as emails to users, send form results, one off sends to contacts, password reset, etc. They are purposively not applied to Email entities as this feature was meant to be used for systems that leverage headers to know what IP to send email from and we do not want an email blast to accidentally go out with a header meant for system email. 

It also fixes setting customer headers for SparkPost and support for contact tokens in headers.

[//]: # ( As applicable: )

#### Steps to test this PR:
1. Run Doctrine migrations `php app/console doctrine:migrations:migrate
2. Create a new or edit an existing email
3. Go to the advanced tab and click add a value button beside custom headers label
4. Add headers as label/value paris
5. Save and send the email
6. View the headers of the received emails and they should be present

Sparkpost headers:
1. Configure Sparkpost as a transport. 
2. Follow the steps above but use a contact field token in the value of a header
3. Check the email that the header is in place and that the value is set as the token value

For global headers:
1. Add custom headers to Configuration -> Email settings
2. Send a broadcast email -> system headers should not be there (only headers defined in the Email's Advanced tab
3. Send an email directly to a contact -> system headers should be present
